Adding batch handling for popping items from RealFIFO

yue9944882 opened the pull request:
/kind feature
/sig api-machinery

### Summary

Ref: https://github.com/kubernetes/kubernetes/issues/130767

The above issue noted a problem that the write lock acquisition from DeltaFIFO's `processDeltas` can have much lower throughput due to read lock contention:

- Read lock attempts were multi-threaded in a classic controller model so it has higher chance to acquire the lock.
- Write lock works in a single-threaded context so at least one read lock will be able to acquire the lock in-between.

This leads to an problem that the distribution between read lock and write lock are almost 1:1 in a large scale busy cluster. 

This PR enables DeltaFIFO to process items in a batch fashion while preserving the in-queue orders, so it can batch multiple item writes (add/update/delete) within a single write lock acquisition.

```release-note
Improved throughput in the real-FIFO queue used by informer/controllers by adding batch handling for processing watch events.
```

### Performance Test

By running the following unit test which simulates heavy watch event stream. It constructs a simple controller with an informer built from in-memory fake watch, then we use 100 concurrent routines to flood watch events to the watch channel:

> https://gist.github.com/yue9944882/f9ee6bbc81373545253b95abc9a65269


<img width="1001" alt="Screenshot 2025-06-11 at 6 55 48 PM" src="https://github.com/user-attachments/assets/f8dc798b-933f-4b41-95a4-041f3b6f8f12" />


- Achieve ~10x write throughput than reads
- Increasing overall controller throughput by ~6x

cartermckinnon commented:
/cc

hakuna-matatah commented:
/cc

mengqiy reviewed (commented):
This is good performance improvement and FMU it still preserves the ordering of events (we wont' run into https://github.com/kubernetes/kubernetes/pull/127631/files)

deads2k commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -583,6 +583,84 @@ func processDeltas(
+						})
+					} else {
+						if err := clientState.Update(obj); err != nil {
+							return err
```
my quick (potentially incorrect?) skim of the code suggests that an error here results in losing every delta after this one.

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -583,6 +583,84 @@ func processDeltas(
+	// from oldest to newest
+	txns := make([]Transaction, 0)
+	callbacks := make([]func(), 0)
+	txnStore, txnSupported := clientState.(TransactionStore)
```
need insights from @deads2k @liggitt

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/store.go:
```diff
@@ -165,6 +179,38 @@ type cache struct {
 
 var _ Store = &cache{}
 
+func (c *cache) Transaction(txns ...Transaction) error {
```
i think 100ms might be too tight for larger batch? added with 500ms

xigang commented:
/cc

yue9944882 commented:
/retest

liggitt commented on staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go:
```diff
@@ -222,28 +232,52 @@ func (i *storeIndex) deleteKeyFromIndex(key, indexValue string, index Index) {
+func (c *threadSafeMap) Transaction(fns ...TransactionFunc) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	unsafe := c.unsafe()
```
Maybe rename skipWriteLock to inLockedTransaction for clarity?

Is there a reason we'd skip just write locks and not read locks?

Do we also need to make sure anyone who gets a handle to the lock-free store cannot retain a handle to it in an unlocked state that outlives the Transaction call, or make asynchronous / multi-threaded calls to the unprotected methods? Exposing an unprotected store to arbitrary injected functions seems ripe for accidents.

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -583,6 +584,72 @@ func processDeltas(
+			}
+		}
+	}
+	defer func() {
```
good catch. removed `defer`, it was not added quite intentionally i think

liggitt commented on staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go:
```diff
@@ -222,28 +232,52 @@ func (i *storeIndex) deleteKeyFromIndex(key, indexValue string, index Index) {
 	index *storeIndex
 }
 
+func (c *threadSafeMap) Transaction(fns ...TransactionFunc) {
```
(or scope down transaction to not be opaque functions and only support explicit add / update / delete operations)

yue9944882 replied:
the unsafe instance is now internal-only in the new rev

yue9944882 commented:
> My findings support this being a big improvement with 1000 a good pick for batch size providing even 30x improvement for write throughput while improving read throughput by 20% 

@serathius we also experimented with some bolder batch size including 10k or even 100k. e.g. with 10k, we only saw roughly 20% improvement by increasing that to 10k in the graph below. but it's probably good to make it a part of the informer constructor option to further customize that. i feel like 1k should be generally good for most clusters.


<img width="1003" height="337" alt="Screenshot 2025-10-24 at 6 02 06 PM" src="https://github.com/user-attachments/assets/bb42161e-02ae-4469-bf67-3b08e64deae0" />

yue9944882 committed:
Adding batch handling for popping items from RealFIFO

Signed-off-by: Min Jin <minkimzz@amazon.com>

Update staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go

optimizing fifo loop

Co-authored-by: Marek Siarkowicz <marek.siarkowicz@protonmail.com>
Signed-off-by: Min Jin <minkimzz@amazon.com>

refactoring PopBatch to accept []Delta

Signed-off-by: Min Jin <minkimzz@amazon.com>

serathius commented on staging/src/k8s.io/client-go/features/known_features.go:
```diff
@@ -75,9 +81,10 @@ const (
+	ClientsAllowCBOR:             {Default: false, PreRelease: Alpha},
+	ClientsPreferCBOR:            {Default: false, PreRelease: Alpha},
+	InOrderInformers:             {Default: true, PreRelease: Beta},
+	InOrderInformersBatchProcess: {Default: false, PreRelease: Alpha},
```
Can we go to Beta default? I think we want to have feature gate as escape hatch, but feature itself can be enabled by default.

serathius commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -585,6 +586,70 @@ func processDeltas(
+			obj := d.Object
+			switch d.Type {
+			case Sync, Replaced, Added, Updated:
+				if old, exists, err := clientState.Get(obj); err == nil && exists {
```
Will it return old correct old if there are two operations for the same object?

serathius replied:
Oh I found below that we avoid batching with same key.

serathius commented:
Please add release note

serathius commented:
/lgtm

liggitt commented on staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go:
```diff
@@ -229,13 +244,41 @@ type threadSafeMap struct {
+	trace := utiltrace.New("ThreadSafeMap Transaction Process",
+		utiltrace.Field{Key: "Size", Value: len(txns)},
+		utiltrace.Field{Key: "Reason", Value: "Slow batch process due to too many items"})
+	defer trace.LogIfLong(500 * time.Millisecond)
```
do we know what is realistic for processing the default batch size of transactions? should this scale with the number of transactions passed in rather than be a fixed time? we don't want this to be so short it ~always logs, or be so long it ~never logs

yue9944882 replied:
Based on our experiement in a 20k node clusterm we only saw two occurrences of this trace log in the past 2 weeks. interestingly, all the log lines happened when the batch size is 1, i believe this delay is caused by GC'ed in a busy process. personally i don't feel a strong difference between time-based or count-base batching here, the difference between them is whether they can be influence by GO GCs. How about we launch with count-based batching and open up options for time-based in the future as part of `RealFIFOOptions`?

```
(REDACTED) trace.go:236] Trace[(REDACTED)]: "ThreadSafeMap Transaction Process" Size:1,Reason:Slow batch process due to too many items (REDACTED)(total time: 645ms):
(REDACTED) trace.go:236] Trace[(REDACTED)]: "ThreadSafeMap Transaction Process" Size:1,Reason:Slow batch process due to too many items (REDACTED) (total time: 604ms):
```

liggitt commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +261,65 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+			utiltrace.Field{Key: "ID", Value: id},
+			utiltrace.Field{Key: "Depth", Value: len(f.items)},
+			utiltrace.Field{Key: "Reason", Value: "slow event handlers blocking the queue"})
+		defer trace.LogIfLong(100 * time.Millisecond)
```
does this need to scale based on the length of ids instead of be fixed to a time geared toward processing a single item?

liggitt commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +261,65 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+		unique.Insert(id)
+	}
+	ids := f.items[0:batchSize]
+	f.items = f.items[batchSize:]
```
this is losing the memory leak fix from RealFIFO#Pop where we zero out the Delta{} in items so a reference to the object is not retained:

```
	item := f.items[0]
	// The underlying array still exists and references this object, so the object will not be garbage collected unless we zero the reference.
	f.items[0] = Delta{}
  ...
```

liggitt commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +261,65 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+		defer trace.LogIfLong(100 * time.Millisecond)
+	}
+
+	// we wrap in Deltas here to be compatible with preview Pop functions and those interpreting the return value.
```
if we make a dedicated function signature for processing batch results, can we avoid needing to rewrite each entry like this and just have a single `Deltas`?

The only place I see PopBatch getting called, we're discarding the returned deltas as well... can maybe truncate this to just return error?

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/store.go:
```diff
@@ -167,6 +207,38 @@ type cache struct {
+	txnStore.Transaction(keyedTxns...)
+	if len(errs) > 0 {
+		return TransactionError{
+			Transactions:      txns,
```
> the caller already has this list, I'm not sure we need to include it in the error response

i think it's better to make the error contain more context, otherwise `TransactionError` will only contain the the indices for the successful/failed items. these indices doesn't explain anything unless we also include the request txn array as a part.

> the Transaction objects will be large, since they embed objects ...

the list is a pointer reference to the array, so IMHO the length of the array doesn't make much difference here?

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/store.go:
```diff
@@ -167,6 +207,38 @@ type cache struct {
+		return errors.New("transaction not supported")
+	}
+	keyedTxns := make([]ThreadSafeStoreTransaction, 0)
+	successfulIndices := make([]int, 0)
```
fixed in the new rev

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +261,65 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+		batchSize++
+		unique.Insert(id)
+	}
+	ids := f.items[0:batchSize]
```
no it shouldn't happen. the only possible scenario is when we encounter error (e.g. key func error) when looping at the 1st element in the batch. in that case, we will return from the function, not triggering the process funcs thereafter.

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -48,9 +49,12 @@ type Config struct {
 	Process ProcessFunc
 
+	// ProcessBatch can process a batch of popped Deltas.
+	ProcessBatch ProcessFunc
```
added a new `ProcessBatchFunc ` in the new rev

serathius commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +261,76 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+	f.items = f.items[batchSize:]
+	if f.initialPopulationCount > 0 {
+		f.initialPopulationCount -= batchSize
+	}
```
Ok, can you apply it?

serathius commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo_test.go:
```diff
@@ -974,3 +978,120 @@ func TestRealFIFO_PopShouldUnblockWhenClosed(t *testing.T) {
+	obj1 := mkFifoObj("foo1", 5)
+	obj2 := mkFifoObj("foo2", 5)
+	obj3 := mkFifoObj("foo3", 5)
+	testCases := []struct {
```
I meant that we should add more test cases specifically for actions. Like:
* "Initial 2 items, update 2 items are split into two batches"
* "3 updates, pop 3 unique items should be in one batch"

yue9944882 replied:
added new cases, can you take another look?

serathius commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo_test.go:
```diff
@@ -974,3 +978,169 @@ func TestRealFIFO_PopShouldUnblockWhenClosed(t *testing.T) {
+				{{{Replaced, obj1}}, {{Replaced, obj2}}},
+				{{{Updated, obj1}}, {{Updated, obj2}}},
+			},
+		},
```
Please ensure we have full coverage of all combinations:
* Splitting caused by non-unique objects Like `[obj1, obj2, obj2, obj3] -> [[obj1, obj2], [obj2, obj3]]`
* Splitting caused of initial vs non-initial objects like `initial: [obj1, obj2], actions: [obj3, obj4] -> [[obj1, obj2], [obj3, obj4]]`
* Splitting casued by batch size `batchSize: 3, [obj1, obj2, obj3, obj4] -> [[obj1, obj2, obj3], [obj4]]`

Please also make sure that for each test scenario we can clearly identify cause of splitting and it cannot be mistaken by other. For example if we expect split by non-unique objects the batch size should be set to very high number.

yue9944882 replied:
now the test cases should cover all combinations with clearer name clarified

aojea commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +261,72 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+
+	isInInitialList := !f.hasSynced_locked()
+	unique := sets.NewString()
+	deltas := make([]Deltas, 0, min(len(f.items), f.batchSize))
```
hmm, I think this will be allocated in the stack 👀

aojea commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -585,6 +605,91 @@ func processDeltas(
+		callback()
+	}
+	return err
+}
```
right now `func (c *cache) Transaction(txns ...Transaction) error {` only returns a no `TransactionError` if the cacheStorage is not ThreadSafeStoreWithTransaction.

The code right now will execute all callbacks and return the error independently, should we make this clear and handle also the error to avoid depending on current implementation and define better the contract?

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +261,72 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+		if err != nil {
+			// close the batch if error happens
+			// TODO: log the error when RealFIFOOptions supports passing klog instance like deprecated DeltaFIO
+			if len(deltas) == 0 {
```
>Not sure I like inconsistent behavior here. If the first item in queue has broken key, it will not be ever popped. But if instead nth element is broken, we will pop n items and ignore the error. This can cause subtle bugs that are hard to debug.

hmm IMHO with the current impl it will always not popping item whether it's 1st or Nth item in the queue. but i think it's a big behavior change from the original non-batch behavior because we used to always pop and the process func deal with the keyFunc error.. i think a better impl is to pop the broken item and not leaving the queue congested..

liggitt commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -48,9 +49,13 @@ type Config struct {
 	Process ProcessFunc
 
+	// ProcessBatch can process a batch of popped Deltas, which should return `TransactionError` if not all items
+	// in the batch were successfully processed.
```
Also document:

```
// For batch processing to be used:
// * The client InOrderInformersBatchProcess feature gate must be enabled
// * ProcessBatch must be non-nil
// * Queue must implement QueueWithBatch
// 
// If any of those are false, Process is used and no batch processing is done.
```

yue9944882 replied:
added

liggitt commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -585,6 +606,95 @@ func processDeltas(
 	return nil
 }
 
+// processDeltasInBatch applies a batch of Deltas to the given Store and
```
```suggestion
// processDeltasInBatch applies a batch of Delta objects to the given Store and
```

liggitt commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -585,6 +606,95 @@ func processDeltas(
+	if !txnSupported {
+		var errs []error
+		for _, delta := range deltasList {
+			if err := processDeltas(handler, clientState, delta, isInInitialList); err != nil {
```
only when falling back do we need to wrap each Delta to Deltas (just like the old non-batch Pop path did):
```suggestion
			if err := processDeltas(handler, clientState, Deltas{delta}, isInInitialList); err != nil {
```

yue9944882 replied:
refactored

liggitt commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -585,6 +606,95 @@ func processDeltas(
+			callbacks[SuccessfulIdx]()
+		}
+		// formatting the error so txns doesn't escape and keeps allocated in the stack.
+		return fmt.Errorf("not all items in the batch successfully processed: %w", txnErr)
```
```
		// formatting the error so txns doesn't escape and keeps allocated in the stack.
```

I don't think this does what you think... %w captures the txnErr variable, which includes transactions - https://github.com/golang/go/blob/00ee1860ce877ec9e2e323d7e48dee083e0902bd/src/fmt/errors.go#L49

yue9944882 replied:
thank you for capturing that.. i updated the code to:

```
		// formatting the error so txns doesn't escape and keeps allocated in the stack.
		return fmt.Errorf("not all items in the batch successfully processed: %s", txnErr.Error())
```

i think it will help prevent escape this time.

liggitt commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +261,72 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+			// still pop the broken item out of queue to be compatible with the non-batch behavior it should be safe
+			// when 1st element is broken, however for Nth broken element, there's possible risk that broken item
+			// still can be processed and broke the uniqueness of the batch unexpectedly.
+			deltas = append(deltas, Deltas{item})
```
```suggestion
			deltas = append(deltas, item)
```

liggitt commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -585,6 +606,95 @@ func processDeltas(
+		// formatting the error so txns doesn't escape and keeps allocated in the stack.
+		return fmt.Errorf("not all items in the batch successfully processed: %w", txnErr)
+	}
+	if err != nil {
```
if this is the only type of error we accept, make the signature return *TransactionError and remove the possibility of a type error here?

yue9944882 replied:
> https://github.com/kubernetes/kubernetes/pull/132240/files#r2488298985

i tried to refactor the signature however we're returning a non-TransactionError if the underlying `cacheStorage` does not support transaction (i.e. not implementing `ThreadSafeStoreWithTransaction` interface). though it looks like the only possible implementation of `cacheStorage` is the `*threadSafeMap` currently, wdyt?

liggitt commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +261,72 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+	// https://github.com/kubernetes/kubernetes/issues/103789
+	if len(f.items) > 10 {
+		id, _ := f.keyOf(deltas[0])
+		trace := utiltrace.New("RealFIFO Pop Process",
```
```suggestion
		trace := utiltrace.New("RealFIFO PopBatch Process",
```

liggitt commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +261,72 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+	if len(f.items) > 10 {
+		id, _ := f.keyOf(deltas[0])
+		trace := utiltrace.New("RealFIFO Pop Process",
+			utiltrace.Field{Key: "ID", Value: id},
```
also add the batch size
```suggestion
			utiltrace.Field{Key: "ID", Value: id},
			utiltrace.Field{Key: "BatchSize", Value: len(deltas)},
```

yue9944882 replied:
added to the trace context (it doesn't show "Outdated" cuz i added on another line)

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +261,72 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+			break
+		}
+		unique.Insert(id)
+		// we wrap in Deltas here to be compatible with preview Pop functions and those interpreting the return value.
```
refactored to use flatlist of Delta in the new commit

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +261,72 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+
+	isInInitialList := !f.hasSynced_locked()
+	unique := sets.NewString()
+	deltas := make([]Deltas, 0, min(len(f.items), f.batchSize))
```
updated

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -94,6 +99,9 @@ type ShouldResyncFunc func() bool
 type ProcessFunc func(obj interface{}, isInInitialList bool) error
 
+// ProcessBatchFunc processes multiple objects in batch.
+type ProcessBatchFunc func(deltas []Deltas, isInInitialList bool) error
```
updated

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/controller.go:
```diff
@@ -585,6 +606,95 @@ func processDeltas(
+	if errors.As(err, &txnErr) {
+		// if txn had error, only execute the callbacks for the successful ones
+		for _, SuccessfulIdx := range txnErr.SuccessfulIndices {
+			callbacks[SuccessfulIdx]()
```
done

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/the_real_fifo.go:
```diff
@@ -254,6 +261,72 @@ func (f *RealFIFO) Pop(process PopProcessFunc) (interface{}, error) {
+	f.items = f.items[len(deltas):]
+
+	// Only log traces if the queue depth is greater than 10 and it takes more than
+	// 100 milliseconds to process one item from the queue.
```
updated

yue9944882 commented on staging/src/k8s.io/client-go/tools/cache/store.go:
```diff
@@ -167,6 +206,37 @@ type cache struct {
+func (c *cache) Transaction(txns ...Transaction) error {
+	txnStore, ok := c.cacheStorage.(ThreadSafeStoreWithTransaction)
+	if !ok {
+		return errors.New("transaction not supported")
```
this line is where non-TransactionErro can be from ^^

liggitt commented on staging/src/k8s.io/client-go/tools/cache/store.go:
```diff
@@ -71,6 +72,44 @@ type Store interface {
+	// Transaction allows multiple operations to occur within a single lock acquisitio to
+	// ensure progress can be made when there is contention. A lock-free view of the store
+	// is passed to the specified function, and the programmer must not utilize it outside
+	// this closure.
```
nit: update godoc to drop the mention of lock-free store, etc, since that's not part of the implementation any more

yue9944882 replied:
ah i missed that. good capture

liggitt commented on staging/src/k8s.io/client-go/tools/cache/store.go:
```diff
@@ -71,6 +72,44 @@ type Store interface {
+
+type TransactionError struct {
+	SuccessfulIndices []int
+	FailedIndices     []int
```
```suggestion
	TotalTransactions int
```

liggitt commented on staging/src/k8s.io/client-go/tools/cache/store.go:
```diff
@@ -71,6 +72,44 @@ type Store interface {
+
+func (t *TransactionError) Error() string {
+	return fmt.Sprintf("failed to execute (%v/%v) transactions failed due to: %v",
+		len(t.FailedIndices), len(t.SuccessfulIndices)+len(t.FailedIndices), t.Errors)
```
```suggestion
	return fmt.Sprintf("failed to execute %d/%d transactions: %v",
		t.TotalTransactions-len(t.SuccessfulIndices), t.TotalTransactions, t.Errors)
```

liggitt commented on staging/src/k8s.io/client-go/tools/cache/store.go:
```diff
@@ -71,6 +72,44 @@ type Store interface {
+}
+
+type TransactionStore interface {
+	// Transaction allows multiple operations to occur within a single lock acquisitio to
```
```suggestion
	// Transaction allows multiple operations to occur within a single lock acquisition to
```

liggitt reviewed (commented):
just the TransactionError return type comment, and a couple typo/doc updates, then lgtm

liggitt commented:
/lgtm
/approve

yue9944882 commented:
>diffing /home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/sample-controller/pkg against freshly generated codegen

looks like it needs another rebase..

pohly commented on staging/src/k8s.io/client-go/tools/cache/shared_informer.go:
```diff
@@ -735,6 +736,12 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}, isInInitialList bool
 	return errors.New("object given as Process argument is not Deltas")
 }
 
+func (s *sharedIndexInformer) HandleBatchDeltas(deltas []Delta, isInInitialList bool) error {
```
There's no reason to export this method, is it? Once a method is exported, we have to be extra careful when changing it because it might break someone.

This went into 1.35, so strictly speaking we cannot fix this anymore.

I'll do it anyway in https://github.com/kubernetes/kubernetes/pull/135395 because I need to inject a logger into this call chain.

/cc @michaelasp
